File: mipush_hook/src/main/java/com/nihility/MethodHooker.java
package com.nihility;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;

import androidx.annotation.NonNull;

import com.xiaomi.network.Fallback;
import com.xiaomi.push.service.MIPushNotificationHelper;
import com.xiaomi.push.service.XMPushService;
import com.xiaomi.xmpush.thrift.PushMetaInfo;
import com.xiaomi.xmpush.thrift.XmPushActionContainer;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.After;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;

@Aspect
public class MethodHooker implements HookedMethodHandler {

    private static @NonNull HookedMethodHandler hookHandler() {
        HookedMethodHandler handler = Dependencies.instance().hookedMethodHandler();
        if (handler != null) {
            return handler;
        }
        return new DefaultHookedMethodHandler();
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.shouldSendBroadcast(..)) && args(pushService, packageName, container, metaInfo)")
    public boolean shouldSendBroadcast(
            final ProceedingJoinPoint joinPoint,
            XMPushService pushService, String packageName, XmPushActionContainer container, PushMetaInfo metaInfo) throws Throwable {
        return hookHandler().shouldSendBroadcast(joinPoint, pushService, packageName, container, metaInfo);
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.postProcessMIPushMessage(..)) && args(pushService, pkgName, payload, newMessageIntent)")
    public void postProcessMIPushMessage(
            final ProceedingJoinPoint joinPoint,
            XMPushService pushService, String pkgName, byte[] payload, Intent newMessageIntent) throws Throwable {
        hookHandler().postProcessMIPushMessage(joinPoint, pushService, pkgName, payload, newMessageIntent);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.ClientEventDispatcher.notifyPacketArrival(..)) && args(pushService, chid, data)")
    public void notifyPacketArrival(final JoinPoint joinPoint,
                                    XMPushService pushService, String chid, Object data) {
        hookHandler().notifyPacketArrival(joinPoint, pushService, chid, data);
    }

    @Override
    @Around("(execution(* com.xiaomi.push.service.XMPushService*.*(..))" +
            " || execution(* com.xiaomi.push.service.PacketSync*.*(..))" +
            " || execution(* com.xiaomi.push.service.ClientEventDispatcher*.*(..))" +
            " || execution(* com.xiaomi.push.service.MIPushEventProcessor*.*(..))" +
            " || execution(* com.xiaomi.push.service.MIPushNotificationHelper*.*(..))" +
            " || execution(* com.xiaomi.push.service.NotificationManagerHelper*.*(..))" +
            ") && !within(is(FinalType))")
    public Object debugLog(final ProceedingJoinPoint joinPoint) throws Throwable {
        return hookHandler().debugLog(joinPoint);
    }

    @Override
    @Before("execution(* com.xiaomi.network.Fallback.getHosts(..)) && target(fallback) && args(usePort)")
    public void logFallback(final JoinPoint joinPoint, Fallback fallback, boolean usePort) {
        hookHandler().logFallback(joinPoint, fallback, usePort);
    }

    @Override
    @Before("execution(* com.xiaomi.mipush.sdk.PushMessageProcessor.processIntent(..)) && args(intent)")
    public void processIntent(final JoinPoint joinPoint, Intent intent) {
        hookHandler().processIntent(joinPoint, intent);
    }

    @Override
    @After("execution(* com.xiaomi.push.service.XMPushService.onCreate(..)) && this(pushService)")
    public void onCreate(final JoinPoint joinPoint, XMPushService pushService) throws Throwable {
        hookHandler().onCreate(joinPoint, pushService);
    }


    @Override
    @Before("execution(* com.xiaomi.push.service.XMPushService.onStartCommand(..))")
    public void onStartCommand(final JoinPoint joinPoint) {
        hookHandler().onStartCommand(joinPoint);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.XMPushService.onStart(..)) && args(intent, startId)")
    public void onStart(final JoinPoint joinPoint, Intent intent, int startId) {
        hookHandler().onStart(joinPoint, intent, startId);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.XMPushService.onBind(..)) && args(intent)")
    public void onBind(final JoinPoint joinPoint, Intent intent) {
        hookHandler().onBind(joinPoint, intent);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.XMPushService.onDestroy(..))")
    public void onDestroy(final JoinPoint joinPoint) {
        hookHandler().onDestroy(joinPoint);
    }

    @Override
    @Before("execution(* com.xiaomi.smack.Connection.setConnectionStatus(..)) && args(newStatus, reason, e)")
    public void setConnectionStatus(final JoinPoint joinPoint,
                                    int newStatus, int reason, Exception e) {
        hookHandler().setConnectionStatus(joinPoint, newStatus, reason, e);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.XMPushService.sendMessage*(..)) && args(intent)")
    public void sendMessage(final JoinPoint joinPoint, Intent intent) {
        hookHandler().sendMessage(joinPoint, intent);
    }

    @Override
    @Before("execution(* com.xiaomi.mipush.sdk.ManifestChecker.checkServices(..)) && args(context, pkgInfo)")
    public void logCheckServices(final JoinPoint joinPoint, PackageInfo pkgInfo) {
        hookHandler().logCheckServices(joinPoint, pkgInfo);
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.buildIntent(..))")
    public Intent buildIntent(final ProceedingJoinPoint joinPoint) throws Throwable {
        return hookHandler().buildIntent(joinPoint);
    }

    @Override
    @Around("execution(* com.nihility.XMPushUtils.packToContainer(..))" +
            "|| execution(* com.xiaomi.push.service.MIPushEventProcessor.buildContainer(..))")
    public XmPushActionContainer buildContainerHook(final ProceedingJoinPoint joinPoint) throws Throwable {
        return hookHandler().buildContainerHook(joinPoint);
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.isIntentAvailable(..))")
    public boolean isIntentAvailable(final ProceedingJoinPoint joinPoint) throws Throwable {
        return hookHandler().isIntentAvailable(joinPoint);
    }

    @Override
    @Before("execution(* com.xiaomi.push.service.MIPushEventProcessor.processMIPushMessage(..)) && args(pushService, decryptedContent, packetBytesLen)")
    public void processMIPushMessage(final JoinPoint joinPoint,
                                     XMPushService pushService, byte[] decryptedContent, long packetBytesLen) {
        hookHandler().processMIPushMessage(joinPoint, pushService, decryptedContent, packetBytesLen);
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MiPushMessageDuplicate.isDuplicateMessage(..)) &&" + "args(pushService, packageName, messageId)")
    public boolean isDuplicateMessage(final ProceedingJoinPoint joinPoint, XMPushService pushService, String packageName, String messageId) throws Throwable {
        return hookHandler().isDuplicateMessage(joinPoint, pushService, packageName, messageId);
    }

    @Override
    @Around("execution(* com.xiaomi.push.service.MIPushNotificationHelper.notifyPushMessage(..)) &&" +
            "args(context, container, decryptedContent)")
    public MIPushNotificationHelper.NotifyPushMessageInfo notifyPushMessage(
            final ProceedingJoinPoint joinPoint, Context context, XmPushActionContainer container, byte[] decryptedContent) throws Throwable {
        return hookHandler().notifyPushMessage(joinPoint, context, container, decryptedContent);
    }

}

File: mipush_hook/src/main/java/com/nihility/MethodHooker.kt
package com.nihility

import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import com.xiaomi.network.Fallback
import com.xiaomi.push.service.MIPushNotificationHelper
import com.xiaomi.push.service.XMPushService
import com.xiaomi.xmpush.thrift.PushMetaInfo
import com.xiaomi.xmpush.thrift.XmPushActionContainer
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.After
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before

@Aspect
class MethodHooker : HookedMethodHandler {
    private fun hookHandler(): HookedMethodHandler {
        val handler = Dependencies.instance()?.hookedMethodHandler()
        return handler ?: DefaultHookedMethodHandler()
    }

    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.shouldSendBroadcast(..)) && args(pushService, packageName, container, metaInfo)")
    override fun shouldSendBroadcast(
        joinPoint: ProceedingJoinPoint,
        pushService: XMPushService,
        packageName: String,
        container: XmPushActionContainer,
        metaInfo: PushMetaInfo
    ): Boolean = hookHandler().shouldSendBroadcast(joinPoint, pushService, packageName, container, metaInfo)

    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.postProcessMIPushMessage(..)) && args(pushService, pkgName, payload, newMessageIntent)")
    override fun postProcessMIPushMessage(
        joinPoint: ProceedingJoinPoint,
        pushService: XMPushService,
        pkgName: String,
        payload: ByteArray,
        newMessageIntent: Intent
    ) = hookHandler().postProcessMIPushMessage(joinPoint, pushService, pkgName, payload, newMessageIntent)

    @Before("execution(* com.xiaomi.push.service.ClientEventDispatcher.notifyPacketArrival(..)) && args(pushService, chid, data)")
    override fun notifyPacketArrival(joinPoint: JoinPoint, pushService: XMPushService, chid: String, data: Any) =
        hookHandler().notifyPacketArrival(joinPoint, pushService, chid, data)

    @Around(
        "(execution(* com.xiaomi.push.service.XMPushService*.*(..))" +
            " || execution(* com.xiaomi.push.service.PacketSync*.*(..))" +
            " || execution(* com.xiaomi.push.service.ClientEventDispatcher*.*(..))" +
            " || execution(* com.xiaomi.push.service.MIPushEventProcessor*.*(..))" +
            " || execution(* com.xiaomi.push.service.MIPushNotificationHelper*.*(..))" +
            " || execution(* com.xiaomi.push.service.NotificationManagerHelper*.*(..))" +
            ") && !within(is(FinalType))"
    )
    override fun debugLog(joinPoint: ProceedingJoinPoint): Any? = hookHandler().debugLog(joinPoint)

    @Before("execution(* com.xiaomi.network.Fallback.getHosts(..)) && target(fallback) && args(usePort)")
    override fun logFallback(joinPoint: JoinPoint, fallback: Fallback, usePort: Boolean) =
        hookHandler().logFallback(joinPoint, fallback, usePort)

    @Before("execution(* com.xiaomi.mipush.sdk.PushMessageProcessor.processIntent(..)) && args(intent)")
    override fun processIntent(joinPoint: JoinPoint, intent: Intent) = hookHandler().processIntent(joinPoint, intent)

    @After("execution(* com.xiaomi.push.service.XMPushService.onCreate(..)) && this(pushService)")
    override fun onCreate(joinPoint: JoinPoint, pushService: XMPushService) = hookHandler().onCreate(joinPoint, pushService)

    @Before("execution(* com.xiaomi.push.service.XMPushService.onStartCommand(..))")
    override fun onStartCommand(joinPoint: JoinPoint) = hookHandler().onStartCommand(joinPoint)

    @Before("execution(* com.xiaomi.push.service.XMPushService.onStart(..)) && args(intent, startId)")
    override fun onStart(joinPoint: JoinPoint, intent: Intent, startId: Int) =
        hookHandler().onStart(joinPoint, intent, startId)

    @Before("execution(* com.xiaomi.push.service.XMPushService.onBind(..)) && args(intent)")
    override fun onBind(joinPoint: JoinPoint, intent: Intent) = hookHandler().onBind(joinPoint, intent)

    @Before("execution(* com.xiaomi.push.service.XMPushService.onDestroy(..))")
    override fun onDestroy(joinPoint: JoinPoint) = hookHandler().onDestroy(joinPoint)

    @Before("execution(* com.xiaomi.smack.Connection.setConnectionStatus(..)) && args(newStatus, reason, e)")
    override fun setConnectionStatus(joinPoint: JoinPoint, newStatus: Int, reason: Int, e: Exception) =
        hookHandler().setConnectionStatus(joinPoint, newStatus, reason, e)

    @Before("execution(* com.xiaomi.push.service.XMPushService.sendMessage*(..)) && args(intent)")
    override fun sendMessage(joinPoint: JoinPoint, intent: Intent) = hookHandler().sendMessage(joinPoint, intent)

    @Before("execution(* com.xiaomi.mipush.sdk.ManifestChecker.checkServices(..)) && args(context, pkgInfo)")
    override fun logCheckServices(joinPoint: JoinPoint, pkgInfo: PackageInfo) = hookHandler().logCheckServices(joinPoint, pkgInfo)

    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.buildIntent(..))")
    override fun buildIntent(joinPoint: ProceedingJoinPoint): Intent = hookHandler().buildIntent(joinPoint)

    @Around(
        "execution(* com.nihility.XMPushUtils.packToContainer(..))" +
            "|| execution(* com.xiaomi.push.service.MIPushEventProcessor.buildContainer(..))"
    )
    override fun buildContainerHook(joinPoint: ProceedingJoinPoint): XmPushActionContainer =
        hookHandler().buildContainerHook(joinPoint)

    @Around("execution(* com.xiaomi.push.service.MIPushEventProcessor.isIntentAvailable(..))")
    override fun isIntentAvailable(joinPoint: ProceedingJoinPoint): Boolean = hookHandler().isIntentAvailable(joinPoint)

    @Before("execution(* com.xiaomi.push.service.MIPushEventProcessor.processMIPushMessage(..)) && args(pushService, decryptedContent, packetBytesLen)")
    override fun processMIPushMessage(
        joinPoint: JoinPoint,
        pushService: XMPushService,
        decryptedContent: ByteArray,
        packetBytesLen: Long
    ) = hookHandler().processMIPushMessage(joinPoint, pushService, decryptedContent, packetBytesLen)

    @Around("execution(* com.xiaomi.push.service.MiPushMessageDuplicate.isDuplicateMessage(..)) &&args(pushService, packageName, messageId)")
    override fun isDuplicateMessage(
        joinPoint: ProceedingJoinPoint,
        pushService: XMPushService,
        packageName: String,
        messageId: String
    ): Boolean = hookHandler().isDuplicateMessage(joinPoint, pushService, packageName, messageId)

    @Around("execution(* com.xiaomi.push.service.MIPushNotificationHelper.notifyPushMessage(..)) &&args(context, container, decryptedContent)")
    override fun notifyPushMessage(
        joinPoint: ProceedingJoinPoint,
        context: Context,
        container: XmPushActionContainer,
        decryptedContent: ByteArray
    ): MIPushNotificationHelper.NotifyPushMessageInfo =
        hookHandler().notifyPushMessage(joinPoint, context, container, decryptedContent)
}
